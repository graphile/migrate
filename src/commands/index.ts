export { init } from "./init";
export { migrate } from "./migrate";
export { watch } from "./watch";
export { commit } from "./commit";
export { uncommit } from "./uncommit";
export { status } from "./status";
export { reset } from "./reset";
export { compile } from "./compile";
export { run } from "./run";
