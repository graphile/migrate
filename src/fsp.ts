import * as fs from "fs";
import { promisify } from "util";

export const readFile = promisify(fs.readFile);
export const writeFile = promisify(fs.writeFile);
export const stat = promisify(fs.stat);
export const readdir = promisify(fs.readdir);
export const mkdir = promisify(fs.mkdir);
export const unlink = promisify(fs.unlink);
