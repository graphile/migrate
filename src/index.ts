export * from "./commands";
export { Settings } from "./settings";
export { Logger, LogFunctionFactory, LogLevel } from "@graphile/logger";
export { defaultLogger } from "./logger";
