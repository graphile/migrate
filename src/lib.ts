export function mergeWithoutClobbering(
  original: { [key: string]: string | undefined },
  newStuff: { [key: string]: string | undefined },
  message: string,
): { [key: string]: string | undefined } {
  const result = { ...original };
  for (const key in newStuff) {
    if (typeof result[key] === "undefined" || result[key] === newStuff[key]) {
      result[key] = newStuff[key];
    } else {
      throw new Error(
        `Refusing to clobber '${key}' (from '${original[key]}' to '${newStuff[key]}'): ${message}`,
      );
    }
  }

  return result;
}
