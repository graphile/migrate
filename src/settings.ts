import { parse } from "pg-connection-string";

import {
  ActionSpec,
  CommandActionSpec,
  makeValidateActionCallback,
  SqlActionSpec,
} from "./actions";

export type Actions = string | Array<string | ActionSpec>;

export function isActionSpec(o: unknown): o is ActionSpec {
  if (!(typeof o === "object" && o && typeof o["_"] === "string")) {
    return false;
  }

  // After here it's definitely an action spec; but we should still validate the
  // other properties.

  if ("shadow" in o && typeof o["shadow"] !== "boolean") {
    throw new Error(
      `'${o["_"]}' action has 'shadow' property of type '${typeof o[
        "shadow"
      ]}'; expected 'boolean' (or not set)`,
    );
  }

  return true;
}

export function isSqlActionSpec(o: unknown): o is SqlActionSpec {
  if (!isActionSpec(o) || o._ !== "sql") {
    return false;
  }
  if (typeof o["file"] !== "string") {
    throw new Error("SQL command requires 'file' property to be set");
  }
  return true;
}

export function isCommandActionSpec(o: unknown): o is CommandActionSpec {
  if (!isActionSpec(o) || o._ !== "command") {
    return false;
  }

  // Validations
  if (typeof o["command"] !== "string") {
    throw new Error(
      `Command action has 'command' property of type '${typeof o[
        "command"
      ]}'; expected 'string'`,
    );
  }

  return true;
}

/**
 * This type is not trusted; to use the values within it, it must be
 * parsed/validated into ParsedSettings.
 */
export interface Settings {
  connectionString?: string;
  shadowConnectionString?: string;
  rootConnectionString?: string;
  databaseOwner?: string;
  migrationsFolder?: string;
  manageGraphileMigrateSchema?: boolean;
  pgSettings?: {
    [key: string]: string;
  };
  placeholders?: {
    [key: string]: string;
  };
  afterReset?: Actions;
  afterAllMigrations?: Actions;
  afterCurrent?: Actions;
  blankMigrationContent?: string;
}

// NOTE: only override values that differ (e.g. changing non-nullability)
export interface ParsedSettings extends Settings {
  connectionString: string;
  rootConnectionString: string;
  databaseOwner: string;
  databaseName: string;
  shadowDatabaseName?: string;
  migrationsFolder: string;
  afterReset: ActionSpec[];
  afterAllMigrations: ActionSpec[];
  afterCurrent: ActionSpec[];
  blankMigrationContent: string;
}

export async function parseSettings(
  settings: Settings,
  requireShadow = false,
): Promise<ParsedSettings> {
  if (!settings) {
    throw new Error("Expected settings object");
  }
  if (typeof settings !== "object") {
    throw new Error("Expected settings object, received " + typeof settings);
  }
  const errors: Array<string> = [];
  const keysToCheck = Object.keys(settings);
  const checkedKeys: Array<string> = [];
  async function check<T = void>(
    key: string,
    callback: (key: unknown) => T | Promise<T>,
  ): Promise<T> {
    checkedKeys.push(key);
    const value = settings[key];
    try {
      return await callback(value);
    } catch (e) {
      errors.push(`Setting '${key}': ${e.message}`);
      return void 0 as never;
    }
  }

  const connectionString = await check(
    "connectionString",
    (rawConnectionString = process.env.DATABASE_URL): string => {
      if (typeof rawConnectionString !== "string") {
        throw new Error(
          "Expected a string, or for DATABASE_URL envvar to be set",
        );
      }
      return rawConnectionString;
    },
  );

  const rootConnectionString = await check(
    "rootConnectionString",
    (
      rawRootConnectionString = process.env.ROOT_DATABASE_URL || "template1",
    ): string => {
      if (typeof rawRootConnectionString !== "string") {
        throw new Error(
          "Expected a string, or for ROOT_DATABASE_URL envvar to be set",
        );
      }
      return rawRootConnectionString;
    },
  );

  const migrationsFolder = await check(
    "migrationsFolder",
    (rawMigrationsFolder = `${process.cwd()}/migrations`): string => {
      if (typeof rawMigrationsFolder !== "string") {
        throw new Error("Expected a string");
      }
      return rawMigrationsFolder;
    },
  );

  const blankMigrationContent = await check(
    "blankMigrationContent",
    (rawBlankMigrationContent = "-- Enter migration here\n"): string => {
      if (typeof rawBlankMigrationContent !== "string") {
        throw new Error("Expected a string");
      }
      return rawBlankMigrationContent;
    },
  );

  const { user, database: databaseName } = parse(connectionString || "");
  const databaseOwner = await check(
    "databaseOwner",
    (rawDatabaseOwner = user || databaseName) => {
      if (typeof rawDatabaseOwner !== "string") {
        throw new Error(
          "Expected a string or for user or database name to be specified in connectionString",
        );
      }
      return rawDatabaseOwner;
    },
  );

  const shadowConnectionString = await check(
    "shadowConnectionString",
    (rawShadowConnectionString = process.env.SHADOW_DATABASE_URL) => {
      if (requireShadow) {
        if (typeof rawShadowConnectionString !== "string") {
          throw new Error(
            "Expected `shadowConnectionString` to be a string, or for SHADOW_DATABASE_URL to be set",
          );
        }
        return rawShadowConnectionString;
      }
      return null;
    },
  );
  const { database: shadowDatabaseName } = parse(shadowConnectionString || "");

  await check("pgSettings", pgSettings => {
    if (pgSettings) {
      if (typeof pgSettings !== "object" || pgSettings === null) {
        throw new Error("Expected settings.pgSettings to be an object");
      }
      const badKeys = Object.keys(pgSettings).filter(key => {
        const value = pgSettings[key];
        return typeof value !== "string" && typeof value !== "number";
      });
      if (badKeys.length) {
        throw new Error(
          `Invalid pgSettings for keys '${badKeys.join(
            ", ",
          )}' - expected string` /* Number is acceptable, but prefer string. Boolean not acceptable. */,
        );
      }
    }
  });

  const placeholders = await check("placeholders", (rawPlaceholders):
    | { [key: string]: string }
    | undefined => {
    if (rawPlaceholders) {
      if (typeof rawPlaceholders !== "object" || rawPlaceholders === null) {
        throw new Error("Expected settings.placeholders to be an object");
      }
      const badKeys = Object.keys(rawPlaceholders).filter(
        key => !/^:[A-Z][0-9A-Z_]+$/.exec(key),
      );
      if (badKeys.length) {
        throw new Error(
          `Invalid placeholders keys '${badKeys.join(
            ", ",
          )}' - expected to follow format ':ABCD_EFG_HIJ'`,
        );
      }
      const badValueKeys = Object.keys(rawPlaceholders).filter(key => {
        const value = rawPlaceholders[key];
        return typeof value !== "string";
      });
      if (badValueKeys.length) {
        throw new Error(
          `Invalid placeholders values for keys '${badValueKeys.join(
            ", ",
          )}' - expected string`,
        );
      }
      return Object.entries(rawPlaceholders).reduce(
        (
          memo: { [key: string]: string },
          [key, value],
        ): { [key: string]: string } => {
          if (value === "!ENV") {
            const envvarKey = key.substr(1);
            const envvar = process.env[envvarKey];
            if (!envvar) {
              throw new Error(
                `Could not find environmental variable '${envvarKey}'`,
              );
            }
            memo[key] = envvar;
          }
          return memo;
        },
        { ...rawPlaceholders },
      );
    }
    return undefined;
  });

  const validateAction = makeValidateActionCallback();
  const rootValidateAction = makeValidateActionCallback(true);

  const afterReset = await check("afterReset", rootValidateAction);
  const afterAllMigrations = await check("afterAllMigrations", validateAction);
  const afterCurrent = await check("afterCurrent", validateAction);

  const manageGraphileMigrateSchema = await check(
    "manageGraphileMigrateSchema",
    mgms => {
      const type = typeof mgms;
      if (type !== "undefined" && type !== "boolean") {
        throw new Error(
          `Expected boolean, received '${
            type === "object" && !mgms ? "null" : type
          }'`,
        );
      }
      return mgms !== false;
    },
  );

  /******/

  const uncheckedKeys = keysToCheck.filter(key => !checkedKeys.includes(key));
  if (uncheckedKeys.length) {
    errors.push(
      `The following config settings were not understood: '${uncheckedKeys.join(
        "', '",
      )}'`,
    );
  }

  if (connectionString) {
    if (!databaseOwner) {
      errors.push(
        "Could not determine the database owner, please add the 'databaseOwner' setting.",
      );
    }

    if (!databaseName) {
      errors.push(
        "Could not determine the database name, please ensure connectionString includes the database name.",
      );
    }
  }

  if (requireShadow && !shadowDatabaseName) {
    errors.push(
      "Could not determine the shadow database name, please ensure shadowConnectionString includes the database name.",
    );
  }

  if (errors.length) {
    throw new Error(
      `Errors occurred during settings validation:\n- ${errors.join("\n- ")}`,
    );
  }
  if (!databaseName) {
    // This is just to appease TypeScript, this should be caught above.
    throw new Error("Could not determine databaseName");
  }

  return {
    ...settings,
    afterReset,
    afterAllMigrations,
    afterCurrent,
    rootConnectionString,
    connectionString,
    manageGraphileMigrateSchema,
    databaseOwner,
    migrationsFolder,
    databaseName,
    shadowConnectionString: shadowConnectionString
      ? shadowConnectionString
      : void 0,
    shadowDatabaseName: shadowDatabaseName ? shadowDatabaseName : void 0,
    placeholders,
    blankMigrationContent,
  };
}

/**
 * Overrides the databaseName in rootConnectionString and returns the resulting
 * connection string.
 */
export function makeRootDatabaseConnectionString(
  parsedSettings: ParsedSettings,
  databaseName: string,
): string {
  const { rootConnectionString } = parsedSettings;
  if (!rootConnectionString) {
    throw new Error(
      "Cannot execute SQL as root since rootConnectionString / ROOT_DATABASE_URL is not specified",
    );
  }
  const parsed = parse(rootConnectionString);
  // TODO: factor in other connection parameters
  let str = "postgres://";
  if (parsed.user) {
    str += encodeURIComponent(parsed.user);
  }
  if (parsed.password) {
    str += ":" + encodeURIComponent(parsed.password);
  }
  if (parsed.user || parsed.password) {
    str += "@";
  }
  if (parsed.host) {
    str += parsed.host;
  }
  if (parsed.port) {
    str += ":" + parsed.port;
  }
  str += "/";
  str += databaseName;
  let sep = "?";
  const q = (key: string, val: string | null | undefined | boolean): string => {
    if (val != null) {
      const str =
        sep +
        encodeURIComponent(key) +
        "=" +
        encodeURIComponent(val === true ? "1" : val === false ? "0" : val);
      if (sep === "?") {
        sep = "&";
      }
      return str;
    }
    return "";
  };
  str += q("ssl", parsed.ssl);
  str += q("client_encoding", parsed.client_encoding);
  str += q("application_name", parsed.application_name);
  str += q("fallback_application_name", parsed.fallback_application_name);
  return str;
}
