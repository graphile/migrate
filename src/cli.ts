#!/usr/bin/env node
/* eslint-disable @typescript-eslint/explicit-function-return-type,no-console */
import * as fs from "fs";

import { uncommit } from "./commands/uncommit";
import { commit, migrate, reset, status, watch } from "./index";

function getSettings() {
  let data;
  try {
    data = fs.readFileSync(`${process.cwd()}/.gmrc`, "utf8");
  } catch (e) {
    throw new Error(
      "No .gmrc file found; please run `graphile-migrate init` first.",
    );
  }
  try {
    return JSON.parse(data);
  } catch (e) {
    throw new Error("Failed to parse .gmrc file: " + e.message);
  }
}

async function main() {
  const argv = process.argv.slice(2);
  const [cmd] = argv;
  if (argv.length === 0 || cmd === "migrate") {
    const shadow = argv.includes("--shadow");
    const force = argv.includes("--force");
    await migrate(getSettings(), shadow, force);
  } else if (cmd === "watch") {
    const once = argv.includes("--once");
    const shadow = argv.includes("--shadow");
    await watch(getSettings(), once, shadow);
  } else if (cmd === "reset") {
    const shadow = argv.includes("--shadow");
    await reset(getSettings(), shadow);
  } else if (cmd === "commit") {
    await commit(getSettings());
  } else if (cmd === "uncommit") {
    await uncommit(getSettings());
  } else if (cmd === "status") {
    let exitCode = 0;
    const details = await status(getSettings());
    const remainingCount = details.remainingMigrations.length;
    if (remainingCount) {
      console.log(
        `There are ${remainingCount} committed migrations pending:\n\n  ${details.remainingMigrations.join(
          "\n  ",
        )}`,
      );
      exitCode += 1;
    }
    if (details.hasCurrentMigration) {
      if (exitCode) {
        console.log();
      }
      console.log(
        "The current.sql migration is not empty and has not been committed.",
      );
      exitCode += 2;
    }

    // ESLint false positive.
    // eslint-disable-next-line require-atomic-updates
    process.exitCode = exitCode;

    if (exitCode === 0) {
      console.log("Up to date.");
    }
  } else {
    // eslint-disable-next-line no-console
    console.error(`Command '${cmd || ""}' not understood`);
    process.exit(1);
  }
}

main().catch(e => {
  // eslint-disable-next-line no-console
  console.error(e);
  process.exit(1);
});
