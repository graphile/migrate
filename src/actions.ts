import { withClient } from "./pg";
import {
  ParsedSettings,
  isCommandActionSpec,
  isSqlActionSpec,
  isActionSpec,
} from "./settings";
import { generatePlaceholderReplacement } from "./migration";
import * as fsp from "./fsp";
import { exec as rawExec } from "child_process";
import { promisify } from "util";
import { parse } from "pg-connection-string";

interface ActionSpecBase {
  _: string;
  shadow?: boolean;
}

export interface SqlActionSpec extends ActionSpecBase {
  _: "sql";
  file: string;
}

export interface CommandActionSpec extends ActionSpecBase {
  _: "command";
  command: string;
}

export type ActionSpec = SqlActionSpec | CommandActionSpec;

const exec = promisify(rawExec);

export async function executeActions(
  parsedSettings: ParsedSettings,
  shadow = false,
  actions: ActionSpec[]
): Promise<void> {
  if (!actions) {
    return;
  }
  const connectionString = shadow
    ? parsedSettings.shadowConnectionString
    : parsedSettings.connectionString;
  if (!connectionString) {
    throw new Error(
      "Could not determine connection string for running commands"
    );
  }
  const { database: databaseName, user: databaseUser } = parse(
    connectionString
  );
  if (!databaseName) {
    throw new Error("Could not extract database name from connection string");
  }
  for (const actionSpec of actions) {
    if (actionSpec.shadow !== undefined && actionSpec.shadow !== shadow) {
      continue;
    }
    if (actionSpec._ === "sql") {
      const body = await fsp.readFile(
        `${parsedSettings.migrationsFolder}/${actionSpec.file}`,
        "utf8"
      );
      await withClient(
        connectionString,
        parsedSettings,
        async (pgClient, context) => {
          const query = generatePlaceholderReplacement(
            parsedSettings,
            context
          )(body);
          await pgClient.query({
            text: query,
          });
        }
      );
    } else if (actionSpec._ === "command") {
      // Run the command
      const { stdout, stderr } = await exec(actionSpec.command, {
        env: {
          ...process.env,
          PATH: process.env.PATH,
          DATABASE_URL: connectionString, // DO NOT USE THIS! It can be misleadling.
          GM_DBNAME: databaseName,
          GM_DBUSER: databaseUser,
          GM_DBURL: connectionString,
          ...(shadow
            ? {
                GM_SHADOW: "1",
              }
            : null),
        },
        encoding: "utf8",
        // 50MB of log data should be enough for any reasonable migration... right?
        maxBuffer: 50 * 1024 * 1024,
      });
      if (stdout) {
        // eslint-disable-next-line no-console
        console.log(stdout);
      }
      if (stderr) {
        // eslint-disable-next-line no-console
        console.error(stderr);
      }
    }
  }
}

export function makeValidateActionCallback() {
  return async (inputValue: unknown): Promise<ActionSpec[]> => {
    const specs: ActionSpec[] = [];
    if (inputValue) {
      const rawSpecArray = Array.isArray(inputValue)
        ? inputValue
        : [inputValue];
      for (const trueRawSpec of rawSpecArray) {
        // This fudge is for backwards compatibility with v0.0.3
        const isV003OrBelowCommand =
          typeof trueRawSpec === "object" &&
          trueRawSpec &&
          !trueRawSpec["_"] &&
          typeof trueRawSpec["command"] === "string";
        if (isV003OrBelowCommand) {
          // eslint-disable-next-line no-console
          console.warn(
            "DEPRECATED: graphile-migrate now requires command action specs to have an `_: 'command'` property; we'll back-fill this for now, but please update your configuration"
          );
        }
        const rawSpec = isV003OrBelowCommand
          ? { _: "command", ...trueRawSpec }
          : trueRawSpec;

        if (rawSpec && typeof rawSpec === "string") {
          const sqlSpec: SqlActionSpec = {
            _: "sql",
            file: rawSpec,
          };
          specs.push(sqlSpec);
        } else if (isActionSpec(rawSpec)) {
          if (isSqlActionSpec(rawSpec) || isCommandActionSpec(rawSpec)) {
            specs.push(rawSpec);
          } else {
            throw new Error(
              `Action spec of type '${rawSpec["_"]}' not supported; perhaps you need to upgrade?`
            );
          }
        } else {
          throw new Error(
            `Expected action spec to contain an array of strings or action specs; received '${typeof rawSpec}'`
          );
        }
      }
    }
    return specs;
  };
}
