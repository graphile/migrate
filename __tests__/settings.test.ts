import * as path from "path";

import { ParsedSettings, parseSettings } from "../src/settings";

function sanitise(parsedSettings: ParsedSettings) {
  parsedSettings.migrationsFolder =
    "./" + path.relative(process.cwd(), parsedSettings.migrationsFolder);
}

const exampleConnectionString = "postgres://localhost:5432/dbname?ssl=1";
it("parses basic config", async () => {
  await parseSettings({
    connectionString: exampleConnectionString,
  });
});

it("throws error for missing connection string", async () => {
  await expect(parseSettings({})).rejects.toMatchInlineSnapshot(`
          [Error: Errors occurred during settings validation:
          - Setting 'connectionString': Expected a string, or for DATABASE_URL envvar to be set]
        `);
});

it("throws if shadow attempted but no shadow DB", async () => {
  await expect(
    parseSettings(
      {
        connectionString: exampleConnectionString,
      },
      true,
    ),
  ).rejects.toMatchInlineSnapshot(`
          [Error: Errors occurred during settings validation:
          - Setting 'shadowConnectionString': Expected \`shadowConnectionString\` to be a string, or for SHADOW_DATABASE_URL to be set
          - Could not determine the shadow database name, please ensure shadowConnectionString includes the database name.]
        `);
});

describe("actions", () => {
  it("parses string values into SQL actions", async () => {
    const parsedSettings = await parseSettings({
      connectionString: exampleConnectionString,
      afterReset: "foo.sql",
      afterAllMigrations: ["bar.sql", "baz.sql"],
    });
    expect(parsedSettings.afterReset).toEqual([{ _: "sql", file: "foo.sql" }]);
    expect(parsedSettings.afterAllMigrations).toEqual([
      { _: "sql", file: "bar.sql" },
      { _: "sql", file: "baz.sql" },
    ]);
    sanitise(parsedSettings);
    expect(parsedSettings).toMatchSnapshot();
  });

  it("parses SQL actions", async () => {
    const parsedSettings = await parseSettings({
      connectionString: exampleConnectionString,
      afterReset: "foo.sql",
      afterAllMigrations: [
        { _: "sql", file: "bar.sql" },
        { _: "sql", file: "baz.sql" },
      ],
    });
    expect(parsedSettings.afterReset).toEqual([{ _: "sql", file: "foo.sql" }]);
    expect(parsedSettings.afterAllMigrations).toEqual([
      { _: "sql", file: "bar.sql" },
      { _: "sql", file: "baz.sql" },
    ]);
    sanitise(parsedSettings);
    expect(parsedSettings).toMatchSnapshot();
  });

  it("parses command actions", async () => {
    const parsedSettings = await parseSettings({
      connectionString: exampleConnectionString,
      afterAllMigrations: [
        { _: "command", command: "pg_dump --schema-only" },
        { _: "command", command: "graphile-worker --once" },
      ],
    });
    expect(parsedSettings.afterReset).toEqual([]);
    expect(parsedSettings.afterAllMigrations).toEqual([
      { _: "command", command: "pg_dump --schema-only" },
      { _: "command", command: "graphile-worker --once" },
    ]);
    sanitise(parsedSettings);
    expect(parsedSettings).toMatchSnapshot();
  });

  it("parses mixed actions", async () => {
    const parsedSettings = await parseSettings({
      connectionString: exampleConnectionString,
      afterAllMigrations: [
        "foo.sql",
        { _: "sql", file: "bar.sql" },
        { _: "command", command: "pg_dump --schema-only" },
        { _: "command", command: "graphile-worker --once" },
      ],
    });
    expect(parsedSettings.afterReset).toEqual([]);
    expect(parsedSettings.afterAllMigrations).toEqual([
      { _: "sql", file: "foo.sql" },
      { _: "sql", file: "bar.sql" },
      { _: "command", command: "pg_dump --schema-only" },
      { _: "command", command: "graphile-worker --once" },
    ]);
    sanitise(parsedSettings);
    expect(parsedSettings).toMatchSnapshot();
  });

  it("is backwards-compatible with untagged command specs", async () => {
    const parsedSettings = await parseSettings({
      connectionString: exampleConnectionString,
      afterAllMigrations: [
        "foo.sql",
        { _: "sql", file: "bar.sql" },
        { command: "pg_dump --schema-only" } as any,
        { _: "command", command: "graphile-worker --once" },
      ],
    });
    expect(parsedSettings.afterReset).toEqual([]);
    expect(parsedSettings.afterAllMigrations).toEqual([
      { _: "sql", file: "foo.sql" },
      { _: "sql", file: "bar.sql" },
      { _: "command", command: "pg_dump --schema-only" },
      { _: "command", command: "graphile-worker --once" },
    ]);
    sanitise(parsedSettings);
    expect(parsedSettings).toMatchSnapshot();
  });

  it("throws on unknown action type", async () => {
    await expect(
      parseSettings({
        connectionString: exampleConnectionString,
        afterAllMigrations: [
          "foo.sql",
          { _: "sql", file: "bar.sql" },
          { _: "unknown_value", command: "pg_dump --schema-only" } as any,
          { _: "command", command: "graphile-worker --once" },
        ],
      }),
    ).rejects.toMatchInlineSnapshot(`
            [Error: Errors occurred during settings validation:
            - Setting 'afterAllMigrations': Action spec of type 'unknown_value' not supported; perhaps you need to upgrade?]
          `);
  });
});
