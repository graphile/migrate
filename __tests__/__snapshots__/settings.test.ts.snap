// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions is backwards-compatible with untagged command specs 1`] = `
Object {
  "afterAllMigrations": Array [
    Object {
      "_": "sql",
      "file": "foo.sql",
    },
    Object {
      "_": "sql",
      "file": "bar.sql",
    },
    Object {
      "_": "command",
      "command": "pg_dump --schema-only",
    },
    Object {
      "_": "command",
      "command": "graphile-worker --once",
    },
  ],
  "afterCurrent": Array [],
  "afterReset": Array [],
  "connectionString": "postgres://localhost:5432/dbname?ssl=1",
  "databaseName": "dbname",
  "databaseOwner": "dbname",
  "manageGraphileMigrateSchema": true,
  "migrationsFolder": "./migrations",
  "placeholders": undefined,
  "rootConnectionString": "template1",
  "shadowConnectionString": undefined,
  "shadowDatabaseName": undefined,
}
`;

exports[`actions parses SQL actions 1`] = `
Object {
  "afterAllMigrations": Array [
    Object {
      "_": "sql",
      "file": "bar.sql",
    },
    Object {
      "_": "sql",
      "file": "baz.sql",
    },
  ],
  "afterCurrent": Array [],
  "afterReset": Array [
    Object {
      "_": "sql",
      "file": "foo.sql",
    },
  ],
  "connectionString": "postgres://localhost:5432/dbname?ssl=1",
  "databaseName": "dbname",
  "databaseOwner": "dbname",
  "manageGraphileMigrateSchema": true,
  "migrationsFolder": "./migrations",
  "placeholders": undefined,
  "rootConnectionString": "template1",
  "shadowConnectionString": undefined,
  "shadowDatabaseName": undefined,
}
`;

exports[`actions parses command actions 1`] = `
Object {
  "afterAllMigrations": Array [
    Object {
      "_": "command",
      "command": "pg_dump --schema-only",
    },
    Object {
      "_": "command",
      "command": "graphile-worker --once",
    },
  ],
  "afterCurrent": Array [],
  "afterReset": Array [],
  "connectionString": "postgres://localhost:5432/dbname?ssl=1",
  "databaseName": "dbname",
  "databaseOwner": "dbname",
  "manageGraphileMigrateSchema": true,
  "migrationsFolder": "./migrations",
  "placeholders": undefined,
  "rootConnectionString": "template1",
  "shadowConnectionString": undefined,
  "shadowDatabaseName": undefined,
}
`;

exports[`actions parses mixed actions 1`] = `
Object {
  "afterAllMigrations": Array [
    Object {
      "_": "sql",
      "file": "foo.sql",
    },
    Object {
      "_": "sql",
      "file": "bar.sql",
    },
    Object {
      "_": "command",
      "command": "pg_dump --schema-only",
    },
    Object {
      "_": "command",
      "command": "graphile-worker --once",
    },
  ],
  "afterCurrent": Array [],
  "afterReset": Array [],
  "connectionString": "postgres://localhost:5432/dbname?ssl=1",
  "databaseName": "dbname",
  "databaseOwner": "dbname",
  "manageGraphileMigrateSchema": true,
  "migrationsFolder": "./migrations",
  "placeholders": undefined,
  "rootConnectionString": "template1",
  "shadowConnectionString": undefined,
  "shadowDatabaseName": undefined,
}
`;

exports[`actions parses string values into SQL actions 1`] = `
Object {
  "afterAllMigrations": Array [
    Object {
      "_": "sql",
      "file": "bar.sql",
    },
    Object {
      "_": "sql",
      "file": "baz.sql",
    },
  ],
  "afterCurrent": Array [],
  "afterReset": Array [
    Object {
      "_": "sql",
      "file": "foo.sql",
    },
  ],
  "connectionString": "postgres://localhost:5432/dbname?ssl=1",
  "databaseName": "dbname",
  "databaseOwner": "dbname",
  "manageGraphileMigrateSchema": true,
  "migrationsFolder": "./migrations",
  "placeholders": undefined,
  "rootConnectionString": "template1",
  "shadowConnectionString": undefined,
  "shadowDatabaseName": undefined,
}
`;
